### PR TITLE
Build tools improvement: vite for bundling & migrate tests to TS

### PR DESCRIPTION
**Pull request by fnlctrl**
# Overall changes
The scope of the PR became much larger than I initially expected, so let me explain:

1. Vite for bundling
2. To remove webpack (used by karma), [karma-vite](https://github.com/credred/karma-vite) is added
3. To run tests, some adjustments are made
   - Not sure if all of them are necessary, but I ended up making everything work with less magic so I didn't want to change them back
   1. No test logic was changed, only syntax.
   2. Test ordering doesn't matter now as they're independent from each other.
   3. Tests are migrated to TS, everything is imported from `src` directly (so `test/parchment.ts` is no longer necessary)
   4. Removed old setup script (`parchment.ts` and `setup.js`), added `setupContextBeforeEach()` in `setup.ts`
       - Every test file normally imports `setup.ts` and calls `setupContextBeforeEach` function
       - This is less magic than previously binding everything globally or to `this`, and helps TS inference because `ctx` is natually in scope
       - `test/registry` files are changed to simply export everything. The side effects of `TestRegistry.register()` is removed. These calls are aggregated into `getTestRegistry()` inside the new `setup.ts` that ensures a fresh new registry is provided every time.

# Notes
1. Karma happened to be [deprecated](https://github.com/karma-runner/karma/commit/450fdfdac5b999967daec1020f1ac69cf9b854ab) 3 days ago. Funny as I was just thinking of a reason to migrate to `vitest`.
3. However, I couldn't get [vitest browser mode](https://vitest.dev/guide/browser.html) to work on windows. 
5. [Web Test Runner](https://modern-web.dev/docs/test-runner/overview/) and [jasmine-browser-runner](https://github.com/jasmine/jasmine-browser-runner) suggested by Karma's notice didn't work for me either
6. In the end https://github.com/credred/karma-vite worked. 
   - Caveat: it only seems to work on vite <= 4.2.1. 4.3 works but prints a lot of errors. So I pinned vite to "4.2.1"
7. Not sure why `test/unit/block.ts` isn't picked up by GitHub as renamed from `test/unit/block.ts` - tried commiting only renames first and changes later, GitHub still thinks it's delete file then add file
9. The previous PR https://github.com/quilljs/parchment/pull/125 was merged without squashing - I was kind of expecting it so I didn't write serious commit messages, and made back-and-forth changes. Maybe if this one is merged use squashing please? (I really hope GitHub can learn from GitLab to provide the PR author with an option to "Squash commits")

# Future todo
1. Despite all tests passing, there are tons of TS errors in the test code needs fixing
   - I fixed `tagName` for blots only to find out it's already proposed by https://github.com/quilljs/parchment/pull/123/files, so maybe merge that one first?
2. Testing
   - Short term: invesigate why vite 4.3 doesn't work with karma-vite and submit a fix
   - Long term: migrate off karma as it's deprecated
        - `vitest` is the most natural choice as we're already using `vite` but its browser mode is still experimental
        - If being `experimental` doesn't bother you, I can investigate how to make it work on windows.

### CODE REVIEW

**Comment by fnlctrl** [comment]
@luin Would you help to take a look? 
I hope to submit more TS type fixes as we're currently living with 100+ warnings/errors due to importing quill from source😂

**Comment by luin** [comment]
Sure. I thought it was not ready for review. Will take a look soon!

File: karma.conf.js [hunk 1]
```diff
@@ -1,46 +1,16 @@
...
-      stats: {
-        assets: false,
-        chunks: false,
-        colors: true,
-        errorDetails: true,
-        hash: false,
-        timings: false,
-        version: false,
+      {
+        // test ordering doesn't matter
```
Comment (luin) [inline karma.conf.js:7]: Let's remove the comment. We only need a comment if the order matters as that would be counterintuitive.

**Review by luin — COMMENTED** [review]
Overall lgtm!

**Comment by luin** [comment]
> I fixed tagName for blots only to find out it's already proposed by https://github.com/quilljs/parchment/pull/123/files, so maybe merge that one first?

I merged that one so you can rebase onto it. There are some missing cases in that PR like `tagName` can be an array for `BlockBlot` and I saw this PR fixed it so let's keep that in this PR.

> Not sure why test/unit/block.ts isn't picked up by GitHub as renamed from test/unit/block.ts - tried commiting only renames first and changes later, GitHub still thinks it's delete file then add file

Yeah it's weird. Usually a separate commit for renaming should do the trick but I'm okay with that as it's just a test file.

> If being experimental doesn't bother you, I can investigate how to make it work on windows

I'm okay with the experimental flag.

> Maybe if this one is merged use squashing please

Yeah sure but also you can also squash them before submitting them for review.

File: karma.conf.js (hunk 1 above)
Comment (fnlctrl) [reply karma.conf.js]: fixed

File: package.json [hunk 2]
```diff
@@ -12,28 +12,24 @@
...
+    "karma-jasmine": "^5.1.0",
     "karma-sauce-launcher": "^4.3.6",
-    "karma-webpack": "^5.0.0",
+    "karma-vite": "^1.0.4",
     "prettier": "^2.5.1",
-    "ts-loader": "^9.2.8",
     "typescript": "^4.9.5",
-    "webpack": "^5.70.0",
-    "webpack-cli": "^4.9.2"
+    "vite": "4.2.1"
```
[the remark this answers is not in the document]
Comment (fnlctrl) [reply package.json]: fixed

**Comment by luin** [comment]
Awesome! Thanks 👍

**Pull request merged**